### PR TITLE
Allow audit admins to upload jurisdiction files

### PR DESCRIPTION
**Pull request by jonahkagan**
Task: #1439 

Gives audit admins the ability to upload jurisdiction files in the jurisdiction detail modal on the progress screen.

Details of the changes are in each commit.

Demo video (seemed to be too large for Github): https://drive.google.com/file/d/1ovUio7uelbb7Or82N23IQb6RZk8IzowB/view?usp=sharing

### CODE REVIEW

File: client/src/components/useJurisdictions.ts
Comment (arsalansufi) [inline client/src/components/useJurisdictions.ts:139]: Nice find! I _think_ I ran into some similar unexpected fetches while working on batch tally sheets (perhaps explained by other refetch-triggering events)

File: client/src/components/useFileUpload.ts
Comment (arsalansufi) [inline client/src/components/useFileUpload.ts:40]: Worth surfacing the fact that `onFileChange` isn't called the first time a file is loaded in a top-level comment for the hook, since this is unlike most `onChange` handlers (which don't distinguish `empty state --> populated state` from `populated state --> diff populated state`)

This also makes me wonder whether any consumers will ever need to trigger a change on the initial file upload and whether we should plan for that future

File: client/src/components/useFileUpload.ts
Comment (arsalansufi) [inline client/src/components/useFileUpload.ts:118]: Related to the above comment, why is it that we only need to invalidate queries on subsequent file upload and not initial file upload? Is it because the data we invalidate doesn't exist until the first file is uploaded?

File: server/tests/api/test_ballot_manifest.py
Comment (jonahkagan) [reply server/tests/api/test_ballot_manifest.py:137]: Yeah, I think it does give sufficient confident. And also I've been trying to stick more to testing the interface, not the internals.

File: client/src/components/testUtilities.tsx
Comment (jonahkagan) [reply client/src/components/testUtilities.tsx:215]: Good callout, will change

File: client/src/components/useFileUpload.ts
Comment (jonahkagan) [reply client/src/components/useFileUpload.ts:37]: Yeah, I thought about that! I think refetching on window focus could be nice for users in general (imagining if they switch tabs and come back). That being said, it's definitely a nice-to-have and Arlo's been totally fine without it for now. What do you think? I could go either way.

File: client/src/components/useFileUpload.ts
Comment (jonahkagan) [reply client/src/components/useFileUpload.ts:118]: Yeah, this is the really tricky part of this whole thing. There are two things going on here:
1. When the page initially loads, there's a list of jurisdictions loaded (including file upload status for each jurisdiction). The modal used to be populated with data from that request. However, I didn't want to poll that endpoint during file uploads to get the status for a single file, since it can be pretty expensive (though maybe this was premature optimization...). It also seemed more straightforward for the file upload component+hook to be self-contained and only think about one file. So I needed some way to reload the jurisdiction list after a file upload completed. Since file uploads are only "complete" after background processing is finished, it's not as simple as just reloading once the initial upload request completes - we need to wait until processing finishes. So that's why we invalidate the jurisdictions request on file _load_, but not on the first load (since there's no chance a file has changed on the first load).
2. In some audit types, there are two jurisdiction files to be uploaded (e.g. manifest and CVRs for ballot comparison). CVRs depend on manifests, since we need to know the batch names from the manifest in order to validate the CVR. That means if a user changes the manifest, the CVR needs to be reprocessed. We made a product decision a while back to do that automatically on manifest changes, which is nice from a user perspective and tricky technically. When a manifest file finishes processing, the server will automatically start the reprocessing of the existing CVR file. That means the frontend needs to know to start polling for the CVR file status after the manifest finishes processing. This is the other reason we use the `onFileChange` handler to invalidate CVRs (and batch tallies) on manifest file load, but not on the first load.

Confusing stuff! I wish there was a way to make it simpler, and I'm open to any suggestions.

File: server/tests/batch_comparison/test_batch_tallies.py
Comment (jonahkagan) [reply server/tests/batch_comparison/test_batch_tallies.py:182]: Oh wait, actually JAs don't have access to the download link (it's just not part of the JA UI for file uploads currently), so that's why the test is this way. I think they could definitely have access in the future if we upgrade their UI to match the new AA UI, but that's for later

File: server/tests/api/test_ballot_manifest.py
Comment (arsalansufi) [reply server/tests/api/test_ballot_manifest.py:137]: > I've been trying to stick more to testing the interface, not the internal

:100:

File: client/src/components/Atoms/FileUpload.tsx
Comment (arsalansufi) [reply client/src/components/Atoms/FileUpload.tsx:34]: Got it, that makes sense!

Of course out of scope for this PR but I do like having explicit loading state indicators to avoid a jitter as items disappear and reappear

File: client/src/components/useFileUpload.ts
Comment (arsalansufi) [reply client/src/components/useFileUpload.ts:37]: Good point about refetching when users switch tabs and come back!

I also don't have a strong preference here but like the idea of making `refetchOnWindowFocus: false` the default to avoid confusion when writing tests involving interactions that trigger window focus (e.g. "Why is this fetch call happening twice when I think it should be happening once? Has my code negatively impacted client-side caching?")

If we do do this, I realize some (many?) existing tests may require updating so feel free to defer to a follow-up issue!

File: client/src/components/useFileUpload.ts
Comment (arsalansufi) [reply client/src/components/useFileUpload.ts:40]: Ah! Just meant noting in a function comment so someone doesn't have to read through the implementation of the hook to understand this behavior

File: client/src/components/useFileUpload.ts
Comment (arsalansufi) [reply client/src/components/useFileUpload.ts:118]: Wow, thank you for all this context!

I don't have any immediate thoughts on how to best simplify the larger architecture, but I do still think that it's a bit unexpected from an interface perspective for `onFileChange` to only run on subsequent file changes and not the first.

One alternative may be to call the option `onUploadSuccess` and expose a param indicating whether it's the initial upload, which consumers can then use

```ts
onUploadSuccess: (isInitialUpload?: boolean) => void
```

File: client/src/components/useFileUpload.ts
Comment (arsalansufi) [reply client/src/components/useFileUpload.ts:40]: (Though this may become irrelevant given my idea below)

**Review by arsalansufi — APPROVED** [review]
I know we're still hashing through some longer-term questions, but code looks good and my questions have been answered! LGTM!

File: client/src/components/useFileUpload.ts
Comment (jonahkagan) [reply client/src/components/useFileUpload.ts:118]: I meant `onFileChange` to mean "when a file is either successfully uploaded, uploaded but fails processing, or is deleted" - in other words, whenever the state of the uploaded file in the database changes. That's why it doesn't run on first load, because the uploaded file didn't change. I'm pretty confident that's the behavior we want. Would making that more clear (through name or comment) help?

File: client/src/components/useFileUpload.ts
Comment (jonahkagan) [reply client/src/components/useFileUpload.ts:37]: I like that reasoning. I'll try making it the default and see how it goes

Comment (jonahkagan) [reply client/src/components/useFileUpload.ts]: Ok it was pretty easy - just had to change one test

File: client/src/components/useFileUpload.ts
Comment (arsalansufi) [reply client/src/components/useFileUpload.ts:114]: Ah I see now! The source of my confusion was a misinterpretation of the phrase "first load." I was interpreting it as the first upload when it's actually referring to loading of a previously uploaded file (if any)

Tweaking the existing explanatory comment could be helpful.

If I'm still missing something (very possible haha), happy to hop on a quick chat!

File: client/src/components/useFileUpload.ts
Comment (arsalansufi) [inline client/src/components/useFileUpload.ts:32]: 💯

**Review by arsalansufi — APPROVED** [review]
Changes LGTM!

**Pull request merged**
